### PR TITLE
Allow filtering by exact value.

### PR DESCRIPTION
**Pull request by dakorpar**
Currently all values are by default filtered with LIKE '%value%', this allows to check for exact values when you provide exact = true on addFilterText
I edited all Datasources as well, but I tested only DibiFluentDatasource, don't have ability to test others.

### CODE REVIEW

**Comment by dakorpar** [comment]
@paveljanda can you please provide some input rather or not this will be accepted? If it's going to be rejected I have to create different approach...

**Comment by paveljanda** [comment]
Sorry. I was releasing new version, so I had no time to review this pull request.

Can you please remove the parameter from `::addFilterXX` method and place the functionality into separate methods?

**Comment by paveljanda** [comment]
And yes, i think that feature is great for datagrid.

**Comment by dakorpar** [comment]
@paveljanda should I create also new filter  (FilterExact), instead of FilterText, or you think it's good enough just to use FilterText?

**Comment by paveljanda** [comment]
I would stay with existing `FilterText` and just add a methods `setExact()` and `isExact()` to it.

**Comment by dakorpar** [comment]
@paveljanda it's done, sorry for poor commit message...

File: src/Filter/FilterText.php [hunk 1]
```diff
@@ -23,6 +23,17 @@ class FilterText extends Filter
 	 */
 	protected $type = 'text';
 
+	/**
+	 * @var bool
+	 */
+	protected $exact = FALSE;
+
+	public function __construct($grid, $key, $name, $column)
```
Comment (paveljanda) [inline src/Filter/FilterText.php]: Why is the constructor still there?

**Comment by dakorpar** [comment]
@paveljanda it's removed now, forgot  remove it

**Comment by dakorpar** [comment]
@paveljanda any update?

**Comment by paveljanda** [comment]
Sorry, a have didn't nave much time last weeks. Can you please resolve the conflicts with master branch? Also a would appreciate renaming `FIlterText::setExact()` to `FIlterText::setExactSearch()`. Thanks.

**Comment by paveljanda** [comment]
@dakorpar  Doesn't satisfy you current feature of `FilterText`, the `::setSplitWordsSearch` functionality?

**Comment by dakorpar** [comment]
Not sure, working on other stuff now, so wasn't doing anything related to the grid lately, will let you know as soon as I try.

**Comment by dakorpar** [comment]
@paveljanda sorry for late reply, but no it's not even close, this is still LIKE %value%, and I think it's really important to have = "exact value", I've merged master back to it and refactored FIlterText::setExact() to FIlterText::setExactSearch() and FIlterText::isExact() to FIlterText::isExactSearch()

**Comment by dakorpar** [comment]
@paveljanda any input? I would really like to see this in master, or at least if you see problems that we disscuss them? Sorry for pushing it :)

**Comment by paveljanda** [comment]
I have currently no time to test your PR on all data sources. Can you please add tests for each data source? Thanks a lot! I will merge it then.

**Comment by dakorpar** [comment]
@paveljanda I can write tests, I don't see any database related test yet, how did you plan to setup unit tests for datasources that are connected to database?

**Comment by paveljanda** [comment]
Hmmm, what about using some sqlite file in git repo? Is it appropriate?

**Comment by dakorpar** [comment]
could work for some datasources, I'll set something up in next few days...

**Comment by paveljanda** [comment]
Thanks, man!

**Comment by dakorpar** [comment]
@paveljanda now we're only missing tests for  

- ApiDataSorce (this is not implemented in it, same as Split words filter)
- DibiFluentMssqlDataSource 
- NetteDatabaseTableMssqlDataSource
- NextrasDataSource ->does not support sqlite

What to do with api datasource? should I implement this in it? Is there any service against which to test, and who will implement this on that end? How to set up unit test on it?
And what to do with datasources which are for MSSQL particularry, and Nextras which don't have sqlite support?

**Comment by dakorpar** [comment]
@paveljanda all tests are passing, and implemented in all datasources which have tests.
Only missing now is DibiFluentMssqlDataSoruce and ApiDataSource which does not support splitWords search also so I consider it out of date...
So is it ready now for merge or?
I've also merge master to it...

**Comment by juniwalk** [comment]
Hey @dakorpar, please fix indentation problem in [DoctrineDataSource.php#L234](https://github.com/dakorpar/datagrid/blob/244f914f42c32200613e8efeeecd5ed4b7a425b1/src/DataSource/DoctrineDataSource.php#L234).

Also there is new DoctrineCollectionDataSource.php which is missing this new implementation.

**Comment by dakorpar** [comment]
@juniwalk  thanks for info, I've fixed it, although I'm not quite sure why tests passed before implemeting on DoctrineCollectionDataSource? Can you take a look maybe? testFilterExactSearch should failed, since it would return two result? Also indentation is fixed, in PHPStorm I don't notice things like that, it's tabs/spaces issue so it looks differently here, and differently in my IDE...

File: src/DataSource/DoctrineCollectionDataSource.php [hunk 2]
```diff
@@ -173,7 +173,9 @@ public function applyFilterText(Filter\FilterText $filter)
 		$exprs = [];
 
 		foreach ($filter->getCondition() as $column => $value) {
-			if ($filter->hasSplitWordsSearch() === FALSE) {
+			if($filter->isExactSearch()){
+				Criteria::expr()->eq($column, $value);
```
Comment (juniwalk) [inline src/DataSource/DoctrineCollectionDataSource.php]: You can't do this. You need to save the result into `$exprs[]` and call `continue` after it, just like you did in `DoctrineDataSource.php`

**Comment by juniwalk** [comment]
@dakorpar And as for passing tests - doctrine/collections uses filtering a little differently so it wouldn't necessary return more than one result.

File: src/DataSource/DoctrineCollectionDataSource.php (hunk 2 above)
Comment (dakorpar) [inline src/DataSource/DoctrineCollectionDataSource.php]: Ah yeah, sorry wasn't really looking was to fast for myself :)

**Comment by dakorpar** [comment]
@juniwalk Is this desired? IMHO all datasources should return exactly the same result when same data is behind, no matter which datasource is used? I mean if someone decide to change datasource he expects that grid will work exactly the same...

File: src/DataSource/DoctrineCollectionDataSource.php [hunk 3]
```diff
@@ -173,9 +173,12 @@ public function applyFilterText(Filter\FilterText $filter)
 		$exprs = [];
 
 		foreach ($filter->getCondition() as $column => $value) {
-			if($filter->isExactSearch()){
+			if($filter->isExactSearch()) {
 				Criteria::expr()->eq($column, $value);
```
Comment (juniwalk) [inline src/DataSource/DoctrineCollectionDataSource.php]: `$exprs[] = Criteria::expr()->eq($column, $value);`

**Comment by juniwalk** [comment]
@dakorpar I don't know about desired but it is the way `doctrine/collections` work. It is not about the datasource as it is just wrapper around the collections.

File: src/DataSource/DoctrineCollectionDataSource.php (hunk 3 above)
Comment (dakorpar) [inline src/DataSource/DoctrineCollectionDataSource.php]: haha, sorry, first working day after 20 days, Can't believe I wrote that 😆

**Comment by juniwalk** [comment]
I'll let @paveljanda have a look on this before merging.

**Comment by paveljanda** [comment]
@dakorpar Please fix the identation (stick to tabs..)
And .. yes, can you please squash these commit into a single one? ([Here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) is a pretty good tutorial about squashing commits)

Thanks for your work!

**Comment by dakorpar** [comment]
@paveljanda I did it before, but not sure you can see it here, plus there were merges from master into this, so it's really hard to make all these one commit... Plus I'm not sure changes are visible here at all, I would probably have to create new pull request...
Any more info on it?

**Comment by juniwalk** [comment]
Squashes are normally visible here. You've probably done something wrong.

**Comment by dakorpar** [comment]
@juniwalk @paveljanda better? Or you want whole branch to be just one commit?
Not sure how I'll accomplish this because of merges later...
Squash is causing me some issues, so I rather did reset soft 4 commits behind and then just commit changes and push --force...

**Comment by juniwalk** [comment]
@dakorpar All commits into single one.

**Comment by dakorpar** [comment]
@juniwalk @paveljanda it's one commit now...

**Comment by paveljanda** [comment]
@dakorpar It really is. But please, can you fi the identation? Use tabs, please.

**Comment by dakorpar** [comment]
@paveljanda Is it ok now? I went through files manually and run convert indents to tabs on files, didn't really check my settings, this was default in PHPStorm, and I don't really care what's behind, as long as it looks good in my IDE, but I understand it can look ugly in someone elses IDE; or here on github web...

**Comment by dakorpar** [comment]
@paveljanda @juniwalk any chance of merging this before monday, I'm really looking forward to get rid of this pull request :)

File: src/DataSource/DoctrineDataSource.php [hunk 4]
```diff
@@ -224,14 +224,19 @@ public function applyFilterText(Filter\FilterText $filter)
 		$exprs = [];
 
 		foreach ($condition as $column => $value) {
+			$c = $this->checkAliases($column);
+
+		if($filter->isExactSearch()){
```
Comment (juniwalk) [inline src/DataSource/DoctrineDataSource.php]: @dakorpar Fix this one last indentation and I'll merge it. :)

**Pull request closed without merging**